// biointellect-frontend/src/config/brandingConfig.js
/**
 * brandingConfig.js
 * 
 * The central configuration for hospital branding and identity.
 * Reverted to English-only as per user request.
 */

export const brandingConfig = {
    // Identity
    brandName: 'BioIntellect',
    hospitalName: 'Saudi German Hospital',
    tagline: 'Precision Medicine & Hospital Intelligence',
    shortDescription: 'BioIntellect is a state-of-the-art clinical intelligence platform powered by Saudi German Health to provide world-class healthcare delivery.',

    // Contact Info
    contact: {
        phone: '+20 02 26252400',
        emergency: '16259',
        email: 'info.egypt@sghgroup.net',
        address: 'Joseph Tito St, El Nozha, Cairo, Egypt',
        supportEmail: 'sghcare@sghegy.com'
    },

    // Statistics
    stats: {
        availability: '24/7',
        patientsServed: '150k+',
        specializedStaff: '500+',
        uptime: '99.9%'
    },

    // Institutional Links
    links: {
        officialWebsite: 'https://sghgroup.com.eg',
        portalName: 'SGH Patient Portal'
    },

    // Creative Assets
    assets: {
        logo: '/src/images/BioIntellect.png',
        heroBg: 'medical_hero_bg',
        textureBg: 'medical_texture_bg'
    }
}
