// biointellect-frontend/src/pages/CreateDoctor.jsx
import { useState, useEffect } from 'react'
import { motion } from 'framer-motion'
import { useAuth } from '../context/AuthContext'
import { useGeography } from '../hooks/useGeography'
import { TopBar } from '../components/TopBar'
import { InputField } from '../components/InputField'
import { SelectField } from '../components/SelectField'
import SearchableSelect from '../components/SearchableSelect'
import { AnimatedButton } from '../components/AnimatedButton'
import { specialtyOptions } from '../constants/options'
import styles from './CreateDoctor.module.css'

export const CreateDoctor = ({ onBack, onComplete, userRole }) => {
    const { registerDoctor, isLoading, error, clearError } = useAuth()
    const {
        countries,
        regions,
        hospitals,
        selectCountry,
        selectRegion
    } = useGeography()

    const [formData, setFormData] = useState({
        fullName: '',
        email: '',
        password: '',
        confirmPassword: '',
        specialty: '',
        phone: '',
        licenseNumber: '',
        countryId: '',
        countryCode: '',
        countryName: '',
        regionId: '',
        regionName: '',
        hospitalId: ''
    })

    const [validationErrors, setValidationErrors] = useState({})
    const [success, setSuccess] = useState(false)

    // Set default country (Egypt) once countries are loaded
    useEffect(() => {
        if (countries.length > 0 && !formData.countryId) {
            const defaultCountry = countries.find(c => c.country_name === 'Egypt') || countries[0]
            handleInputChange('countryId', defaultCountry.country_id)
            handleInputChange('countryCode', defaultCountry.country_code)
            handleInputChange('countryName', defaultCountry.country_name)
            selectCountry(defaultCountry.country_id)
        }
    }, [countries])

    const onCountryChange = (e) => {
        const selected = countries.find(c => c.country_id === e.target.value)
        handleInputChange('countryId', e.target.value)
        handleInputChange('countryCode', selected?.country_code || '')
        handleInputChange('countryName', selected?.country_name || '')
        selectCountry(e.target.value)
    }

    const onRegionChange = (e) => {
        const selected = regions.find(r => r.region_id === e.target.value)
        handleInputChange('regionId', e.target.value)
        handleInputChange('regionName', selected?.region_name || '')
        selectRegion(e.target.value)
    }

    const handleInputChange = (field, value) => {
        setFormData(prev => ({ ...prev, [field]: value }))
        if (validationErrors[field]) {
            setValidationErrors(prev => ({ ...prev, [field]: '' }))
        }
        clearError()
    }

    const validateForm = () => {
        const errors = {}
        if (!formData.fullName.trim()) errors.fullName = 'Full name is required'
        if (!formData.email.trim()) errors.email = 'Email is required'
        if (!formData.specialty) errors.specialty = 'Specialty selection is required'
        if (!formData.licenseNumber.trim()) errors.licenseNumber = 'License number is required'
        if (!formData.hospitalId) errors.hospitalId = 'Hospital selection is required'

        const passwordRegex = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[!@#$%^&*()_+\-=[\]{};':"\\|,.<>/?`~]).{16,}$/
        if (!formData.password) {
            errors.password = 'Password is required'
        } else if (!passwordRegex.test(formData.password)) {
            errors.password = 'Password must be 16+ chars with mixed cases, digits & symbols'
        }

        if (formData.password !== formData.confirmPassword) {
            errors.confirmPassword = 'Passwords do not match'
        }

        setValidationErrors(errors)
        return Object.keys(errors).length === 0
    }

    const handleSubmit = async (e) => {
        e.preventDefault()
        if (!validateForm()) return

        const result = await registerDoctor(formData)
        if (result.success) {
            setSuccess(true)
        }
    }

    if (success) {
        return (
            <div className={styles.pageWrapper}>
                <TopBar userRole="administrator" onBack={() => {
                    setSuccess(false); setFormData({
                        fullName: '', email: '', password: '', confirmPassword: '', specialty: '', phone: '', licenseNumber: '',
                        countryId: '', countryCode: '', countryName: '', regionId: '', regionName: '', hospitalId: ''
                    })
                }} />
                <div className={styles.container}>
                    <motion.div
                        className={styles.card}
                        initial={{ opacity: 0, scale: 0.95 }}
                        animate={{ opacity: 1, scale: 1 }}
                    >
                        <div className={styles.successView}>
                            <div className={styles.successIcon}>🛡️</div>
                            <h2 className={styles.title}>Account Provisioned</h2>
                            <p className={styles.subtitle}>
                                Doctor <strong>{formData.fullName}</strong> has been successfully registered.
                                <br /><br />
                                <span style={{ color: 'var(--color-primary)', fontWeight: '600' }}>
                                    ⚠️ ACTION REQUIRED: An activation email has been sent. The doctor must confirm their email before they can sign in.
                                </span>
                            </p>
                            <div className={styles.successActions}>
                                <AnimatedButton variant="primary" fullWidth onClick={() => {
                                    setSuccess(false); setFormData({
                                        fullName: '', email: '', password: '', confirmPassword: '', specialty: '', phone: '', licenseNumber: '',
                                        countryId: '', countryCode: '', countryName: '', regionId: '', regionName: '', hospitalId: ''
                                    })
                                }}>Provision Another Staff Member</AnimatedButton>
                                <AnimatedButton variant="secondary" fullWidth onClick={onBack}>Return to Dashboard</AnimatedButton>
                            </div>
                        </div>
                    </motion.div>
                </div>
            </div>
        )
    }

    return (
        <div className={styles.pageWrapper}>
            <TopBar userRole={userRole} onBack={onBack} />
            <div className={styles.container}>
                <motion.div
                    className={styles.card}
                    initial="hidden"
                    animate="visible"
                    variants={{
                        hidden: { opacity: 0, scale: 0.98 },
                        visible: {
                            opacity: 1,
                            scale: 1,
                            transition: {
                                duration: 0.8,
                                ease: [0.22, 1, 0.36, 1],
                                staggerChildren: 0.15
                            }
                        }
                    }}
                >
                    <div className={styles.header}>
                        <h1 className={styles.title}>Provision Medical Staff</h1>
                        <p className={styles.subtitle}>Add a new healthcare practitioner to the clinical network according to system schema.</p>
                    </div>

                    {error && <div className={styles.alertError}>{error}</div>}

                    <form onSubmit={handleSubmit} className={styles.form}>
                        <div className={styles.grid}>
                            <InputField
                                label="Professional Full Name"
                                placeholder="Dr. Ahmed Ali"
                                value={formData.fullName}
                                onChange={(e) => handleInputChange('fullName', e.target.value)}
                                error={validationErrors.fullName}
                                required
                                autoComplete="name"
                            />
                            <InputField
                                label="Medical Email"
                                type="email"
                                placeholder="a.ali@biointellect.com"
                                value={formData.email}
                                onChange={(e) => handleInputChange('email', e.target.value)}
                                error={validationErrors.email}
                                required
                                autoComplete="username"
                            />
                            <SearchableSelect
                                label="Specialty / System Role"
                                value={formData.specialty}
                                onChange={(e) => handleInputChange('specialty', e.target.value)}
                                options={specialtyOptions}
                                required
                                placeholder="Search medical specialty..."
                            />
                            <InputField
                                label="Medical License Number"
                                placeholder="LIC-XXXXXX"
                                value={formData.licenseNumber}
                                onChange={(e) => handleInputChange('licenseNumber', e.target.value)}
                                error={validationErrors.licenseNumber}
                                required
                            />
                            <SearchableSelect
                                label="Country"
                                value={formData.countryId}
                                onChange={onCountryChange}
                                options={countries.map(c => ({
                                    value: c.country_id,
                                    label: c.country_name,
                                    code: c.country_code
                                }))}
                                required
                                isCountry={true}
                                placeholder="Search country..."
                            />
                            <SearchableSelect
                                label="Region / State"
                                value={formData.regionId}
                                onChange={onRegionChange}
                                options={regions.map(r => ({ value: r.region_id, label: r.region_name }))}
                                required
                                disabled={!formData.countryId}
                                placeholder={!formData.countryId ? "Select country first" : "Search region..."}
                            />
                            <SearchableSelect
                                label="Clinical Hospital"
                                value={formData.hospitalId}
                                onChange={(e) => handleInputChange('hospitalId', e.target.value)}
                                options={hospitals.map(h => ({ value: h.hospital_id, label: h.hospital_name }))}
                                required
                                disabled={!formData.regionId}
                                error={validationErrors.hospitalId}
                                placeholder={!formData.regionId ? "Select region first" : "Search hospitals..."}
                            />
                            <InputField
                                label="Security Password"
                                type="password"
                                placeholder="••••••••••••••••"
                                value={formData.password}
                                onChange={(e) => handleInputChange('password', e.target.value)}
                                error={validationErrors.password}
                                required
                                autoComplete="new-password"
                            />
                            <InputField
                                label="Confirm Password"
                                type="password"
                                placeholder="••••••••••••••••"
                                value={formData.confirmPassword}
                                onChange={(e) => handleInputChange('confirmPassword', e.target.value)}
                                error={validationErrors.confirmPassword}
                                required
                                autoComplete="new-password"
                            />
                            <div className={styles.fullGrid}>
                                <InputField
                                    label="Phone Number"
                                    placeholder="+20 1XX XXX XXXX"
                                    value={formData.phone}
                                    onChange={(e) => handleInputChange('phone', e.target.value)}
                                />
                            </div>
                        </div>

                        <AnimatedButton
                            type="submit"
                            variant="primary"
                            size="large"
                            fullWidth
                            isLoading={isLoading}
                        >
                            Provision Account & Grant Access
                        </AnimatedButton>
                    </form>
                </motion.div>
            </div>
        </div>
    )
}
